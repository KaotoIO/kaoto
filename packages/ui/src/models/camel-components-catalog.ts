import { CatalogKind } from './catalog-kind';

export interface ICamelComponentDefinition {
  component: ICamelComponent;
  componentProperties: ICamelComponentProperties;
  headers: ICamelComponentHeaders;
  properties: ICamelComponentProperties;
}

export interface ICamelComponent {
  kind: CatalogKind.Component;
  name: string;
  title: string;
  description: string;
  deprecated: boolean;
  firstVersion?: string;
  label: string;
  javaType?: string;
  supportLevel?: string;
  groupId?: string;
  artifactId?: string;
  version: string;
  scheme?: string;
  extendsScheme?: string;
  syntax?: string;
  async?: boolean;
  api?: boolean;
  consumerOnly?: boolean;
  producerOnly?: boolean;
  lenientProperties?: boolean;
}

export interface CatalogItemCommon {
  index: number;
  kind: string;
  displayName: string;
  group: string;
  label: string;
  required: boolean;
  javaType: string;
  deprecated: boolean;
  deprecationNote: string;
  autowired: boolean;
  secret: boolean;
  description: string;
}

export interface ICamelComponentProperties {
  type: string;
}

export interface ICamelComponentHeaders extends CatalogItemCommon {
  constantName: string;
}
