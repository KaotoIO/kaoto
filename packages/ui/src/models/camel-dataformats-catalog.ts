import { ICamelProcessorModel, ICamelProcessorProperty } from './camel-processors-catalog';
import { KaotoSchemaDefinition } from './kaoto-schema';

export interface ICamelDataformatDefinition {
  model: ICamelDataformatModel;
  properties: Record<string, ICamelDataformatProperty>;
  propertiesSchema: KaotoSchemaDefinition['schema'];
}

export type ICamelDataformatModel = ICamelProcessorModel;

export type ICamelDataformatProperty = ICamelProcessorProperty;
