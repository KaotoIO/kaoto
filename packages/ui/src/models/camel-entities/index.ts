export * from './base-entity';

/** Camel Route definitions */
export * from './flows/route';
