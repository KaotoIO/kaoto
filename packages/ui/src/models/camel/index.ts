export * from './camel-resource';
export * from './camel-route-resource';
export * from './integration-resource';
export * from './kamelet-resource';
export * from './kamelet-binding-resource';
export * from './pipe-resource';
export * from './source-schema-config';
export * from './source-schema-type';
export { CAMEL_K_K8S_API_VERSION_V1 } from './camel-k-resource';
