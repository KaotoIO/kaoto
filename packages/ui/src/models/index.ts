export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
