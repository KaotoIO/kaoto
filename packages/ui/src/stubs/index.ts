export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
