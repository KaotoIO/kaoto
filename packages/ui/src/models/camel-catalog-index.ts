import { ICamelComponentDefinition } from './camel-components-catalog';
import { ICamelProcessorDefinition } from './camel-processors-catalog';
import { CatalogKind } from './catalog-kind';
import { IKameletDefinition } from './kamelets-catalog';

export interface CamelCatalogIndex {
  catalogs: Catalogs;
  schemas: CatalogEntry[];
}

export interface Catalogs {
  models: CatalogEntry;
  components: CatalogEntry;
  languages: CatalogEntry;
  dataformats: CatalogEntry;
  kamelets: CatalogEntry;
}

export interface CatalogEntry {
  name: string;
  version: string;
  file: string;
}

export type CatalogTypes = Record<string, ICamelComponentDefinition | ICamelProcessorDefinition | IKameletDefinition>;

export interface ComponentsCatalog {
  [CatalogKind.Component]?: Record<string, ICamelComponentDefinition>;
  [CatalogKind.Processor]?: Record<string, ICamelProcessorDefinition>;
  [CatalogKind.Kamelet]?: Record<string, IKameletDefinition>;
  [key: string]: unknown;
}

export interface CatalogCamelComponent {
  type: CatalogKind.Component;
  definition: ICamelComponentDefinition;
}

export interface CatalogCamelProcessor {
  type: CatalogKind.Processor;
  definition: ICamelProcessorDefinition;
}

export interface CatalogKamelet {
  type: CatalogKind.Kamelet;
  definition: IKameletDefinition;
}
