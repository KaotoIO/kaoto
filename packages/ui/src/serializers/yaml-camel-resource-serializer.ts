import { CamelResource } from '../models/camel';
import { parse, stringify } from 'yaml';
import { CamelResourceSerializer } from './camel-resource-serializer';
import { CamelYamlDsl, Integration, Kamelet, KameletBinding, Pipe } from '@kaoto/camel-catalog/types';

export class YamlCamelResourceSerializer implements CamelResourceSerializer {
  /**
   * Regular expression to match commented lines, regardless of indentation
   * Given the following examples, the regular expression should match the comments:
   * ```
   * # This is a comment
   *     # This is an indented comment
   *# This is an indented comment
   * ```
   * The regular expression should match the first three lines
   */
  static readonly COMMENTED_LINES_REGEXP = /^\s*#.*$/;
  comments: string[] = [];

  static isApplicable(_code: unknown): boolean {
    //TODO
    // return !isXML(code);

    return true;
  }

  parse(code: string): CamelYamlDsl | Integration | Kamelet | KameletBinding | Pipe {
    if (!code || typeof code !== 'string') return [];

    this.comments = this.parseComments(code);
    const json = parse(code);
    return json;
  }

  serialize(resource: CamelResource): string {
    let code = stringify(resource, { sortMapEntries: resource.sortFn, schema: 'yaml-1.1' }) || '';
    if (this.comments.length > 0) {
      const comments = this.comments.join('\n');
      code = comments + '\n' + code;
    }
    return code;
  }

  getComments(): string[] {
    return this.comments;
  }

  setComments(comments: string[]): void {
    this.comments = comments;
  }

  private parseComments(code: string): string[] {
    const lines = code.split('\n');
    const comments: string[] = [];
    for (const line of lines) {
      if (line.trim() === '' || YamlCamelResourceSerializer.COMMENTED_LINES_REGEXP.test(line)) {
        comments.push(line);
      } else {
        break;
      }
    }
    return comments;
  }
}
