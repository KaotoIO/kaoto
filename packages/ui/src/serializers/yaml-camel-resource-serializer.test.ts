import { YamlCamelResourceSerializer } from './yaml-camel-resource-serializer';
import { camelRouteJson, camelRouteYaml } from '../stubs';
import { CamelRouteResource } from '../models/camel';

describe('YamlCamelResourceSerializer', () => {
  let serializer: YamlCamelResourceSerializer;

  beforeEach(() => {
    serializer = new YamlCamelResourceSerializer();
  });

  it('parses YAML code into JSON object', () => {
    const result = serializer.parse(camelRouteYaml);
    expect(result).toEqual([camelRouteJson]);
  });

  it('returns empty array for empty or non-string input in parse', () => {
    expect(serializer.parse('')).toEqual([]);
    expect(serializer.parse(null as unknown as string)).toEqual([]);
    expect(serializer.parse(123 as unknown as string)).toEqual([]);
  });

  it('includes comments in serialized YAML string', () => {
    const entities = serializer.parse('# comment1\n' + camelRouteYaml);
    console.log(serializer.comments);
    expect(serializer.comments.includes('# comment1')).toBeTruthy();

    serializer.comments.push('# Comment2');
    const result = serializer.serialize(new CamelRouteResource(entities));
    expect(result).toContain('# Comment2');
  });
});
