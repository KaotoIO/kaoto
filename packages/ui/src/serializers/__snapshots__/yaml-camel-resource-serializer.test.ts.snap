// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YamlCamelResourceSerializer includes comments in  YAML string 1`] = `
"# comment1

# Comment2
- route:
    id: route-8888
    from:
      uri: timer
      parameters:
        timerName: tutorial
      steps:
        - set-header:
            name: myChoice
            simple: \${random(2)}
        - choice:
            otherwise:
              steps:
                - to:
                    uri: "amqp:queue:"
                - to:
                    uri: "amqp:queue:"
                - log:
                    id: log-2
                    message: We got a \${body}
            when:
              - steps:
                  - log:
                      id: log-1
                      message: We got a one.
                simple: \${header.myChoice} == 1
        - to:
            uri: direct:my-route
            parameters:
              bridgeErrorHandler: true
"
`;
